### PR TITLE
Add Service Import and Service Export metrics

### PR DESCRIPTION
**Pull request by maayanf24**
Related issue #349

Signed-off-by: Maayan Friedman <maafried@redhat.com>

### CODE REVIEW

**Comment by maayanf24** [comment]
@skitt @tpantelis @mangelajo @sridhargaddam @nyechiel  - what do you think about prefix the metrics with "submariner_" for the lighthouse?
Should it be prefixed with "lighthouse_" instead?

**Comment by nyechiel** [comment]
> @skitt @tpantelis @mangelajo @sridhargaddam @nyechiel - what do you think about prefix the metrics with "submariner_" for the lighthouse?
> Should it be prefixed with "lighthouse_" instead?

For context, see https://github.com/submariner-io/submariner/pull/1143

Either option should work fine. Technically, users can run Lighthouse without Submariner, but I think that the majority will run both (and even if not, Lighthouse is still a repo under Submariner) - so for simplicity I am leaning towards "submariner_".

**Comment by tpantelis** [comment]
> > @skitt @tpantelis @mangelajo @sridhargaddam @nyechiel - what do you think about prefix the metrics with "submariner_" for the lighthouse?
> > Should it be prefixed with "lighthouse_" instead?
> 
> For context, see [submariner-io/submariner#1143](https://github.com/submariner-io/submariner/pull/1143)
> 
> Either option should work fine. Technically, users can run Lighthouse without Submariner, but I think that the majority will run both (and even if not, Lighthouse is still a repo under Submariner) - so for simplicity I am leaning towards "submariner_".

I agree with "_submariner__" - "_lighthouse_" is just a repo name.  If we were to separate it from submariner then I'd suggest "_service_discovery_".

**Comment by aswinsuryan** [comment]
> > > @skitt @tpantelis @mangelajo @sridhargaddam @nyechiel - what do you think about prefix the metrics with "submariner_" for the lighthouse?
> > > Should it be prefixed with "lighthouse_" instead?
> > 
> > 
> > For context, see [submariner-io/submariner#1143](https://github.com/submariner-io/submariner/pull/1143)
> > Either option should work fine. Technically, users can run Lighthouse without Submariner, but I think that the majority will run both (and even if not, Lighthouse is still a repo under Submariner) - so for simplicity I am leaning towards "submariner_".
> 
> I agree with "_submariner__" - "_lighthouse_" is just a repo name. If we were to separate it from submariner then I'd suggest "_service_discovery_".

+1

**Review by skitt — APPROVED** [review]

**Comment by vthapar** [comment]
@skitt @aswinsuryan @maayanf24 Changes look good but I have a question. Should ServiceExport/Import of same service in different clusters be considered as single or multiple?

File: pkg/agent/controller/agent_test.go [hunk 1]
```diff
@@ -512,7 +514,20 @@ func (c *cluster) init(syncerConfig broker.SyncerConfig) {
...
+
+	bigint, err = rand.Int(rand.Reader, big.NewInt(1000000))
+	Expect(err).To(Succeed())
+
+	serviceExportCounterName := "submariner_service_export" + bigint.String()
+
+	agentController, err := controller.New(&c.agentSpec, syncerConfig, c.localKubeClient,
+		controller.AgentConfig{
+			ServiceImportCounterName: serviceImportCounterName,
+			ServiceExportCounterName: serviceExportCounterName})
```
Comment (tpantelis) [inline pkg/agent/controller/agent_test.go:530]: Can we verify that the metrics are updated where appropriate?

**Comment by maayanf24** [comment]
> @skitt @aswinsuryan @maayanf24 Changes look good but I have a question. Should ServiceExport/Import of same service in different clusters be considered as single or multiple?

@vthapar - each cluster only show local perspective of ServiceExport/Import

File: pkg/agent/controller/agent_test.go (hunk 1 above)
Comment (mangelajo) [reply pkg/agent/controller/agent_test.go:530]: I believe this patch does not bump those metrics, at least I don't see the code

**Review by mangelajo — COMMENTED** [review]
> > @skitt @tpantelis @mangelajo @sridhargaddam @nyechiel - what do you think about prefix the metrics with "submariner_" for the lighthouse?
> > Should it be prefixed with "lighthouse_" instead?
> 
> For context, see [submariner-io/submariner#1143](https://github.com/submariner-io/submariner/pull/1143)
> 
> Either option should work fine. Technically, users can run Lighthouse without Submariner, but I think that the majority will run both (and even if not, Lighthouse is still a repo under Submariner) - so for simplicity I am leaning towards "submariner_".

I had exactly the same train of thought ':D

File: pkg/agent/controller/agent_test.go (hunk 1 above)
Comment (skitt) [reply pkg/agent/controller/agent_test.go:530]: They should be updated by the syncer (in Admiral).

Comment (tpantelis) [reply pkg/agent/controller/agent_test.go:530]: true but for white box testing we're not supposed to know those details. Regardless, adding some counter verification would at least verify that the admiral components were setup correctly. We could also do this in E2E instead. But we could do it in another PR if folks think it's worth it.

**Comment by mangelajo** [comment]
Oh I see it does actually use those prometheus counters from the admiral resource syncer, awesome.

**Review by mangelajo — APPROVED** [review]

**Comment by mangelajo** [comment]
Hey @maayanf24 sorry, this needs a rebase due to conflicts at this point.

**Review by tpantelis — APPROVED** [review]

**Pull request merged**